src/main/java/com/github/cypher/sdk/Client.java

```diff
@@ -1,13 +1,242 @@
+	public void removeAccountDataListener(MapChangeListener<String, String> listener) { accountData.removeListener(listener); }
+
+	private ObservableMap<String, Room> joinRooms   = new ObservableMapWrapper<>(new HashMap<>());
+	private ObservableMap<String, Room> inviteRooms = new ObservableMapWrapper<>(new HashMap<>());
+	private ObservableMap<String, Room> leaveRooms  = new ObservableMapWrapper<>(new HashMap<>());
+
+	public void addJoinRoomsListener     (MapChangeListener<String, Room> listener) { joinRooms.addListener(listener);      }
+	public void removeJoinRoomsListener  (MapChangeListener<String, Room> listener) { joinRooms.removeListener(listener);   }
+	public void addInviteRoomsListener   (MapChangeListener<String, Room> listener) { inviteRooms.addListener(listener);    }
+	public void removeInviteRoomsListener(MapChangeListener<String, Room> listener) { inviteRooms.removeListener(listener); }
+	public void addLeaveRoomsListener    (MapChangeListener<String, Room> listener) { leaveRooms.addListener(listener);     }
+	public void removeLeaveRoomsListener (MapChangeListener<String, Room> listener) { leaveRooms.removeListener(listener);  }
+
+	/**
+	 * @see com.github.cypher.sdk.api.ApiLayer
+	 * @throws IOException
+	 */
+	public Client(ApiLayer api) {
+		this.api = api;
+	}
+
+	/**
+	 * This object returns an existing User object if possible,
+	 * otherwise it creates and caches a new one.
+	 * @param userId The unique ID of the user (e.g. "@morpheus:matrix.org")
+	 * @return A User object
+	 */
+	public User getUser(String userId) {
+		if(users.containsKey(userId)) {
+			return users.get(userId);
+		}
+		User user = new User(api, userId);
+		users.put(userId, user);
+		return user;
+	}
+
+	private Room getOrCreateRoom(Map<String, Room> map, String roomId) {
+		if(map.containsKey(roomId)) {
+			return map.get(roomId);
+		}
+		Room room = new Room(api, this, roomId);
+		map.put(roomId, room);
+		return room;
+	}
+
+	/**
+	 * Call ApiLayer.login(...)
+	 * @see com.github.cypher.sdk.api.ApiLayer#login(String, String, String)
+	 * @throws RestfulHTTPException
+	 * @throws IOException
+	 */
+	public void login(String username, String password, String homeserver) throws RestfulHTTPException, IOException {
+		api.login(username, password, homeserver);
+	}
+
+	/**
+	 * Call ApiLayer.sync(...) and parse the returned data.
+	 * <p>Presence-data is used to update the map of users: {@link #getUser(String)}</p>
+	 * <p>Join-data is used to update the map of rooms the user has joined: {@link #getJoinRooms()}.</p>
+	 * <p>All room maps are observable using the various add*RoomsListener(...) methods</p>
+	 * @see com.github.cypher.sdk.api.ApiLayer#sync(String, String, boolean, User.Presence, int)
+	 * @throws RestfulHTTPException
+	 * @throws IOException
+	 */
+	public void update(int timeout) throws RestfulHTTPException, IOException {
```

Comment:
Would like it if you could break this method down into a few smaller functions with descriptive names, making the code easier to follow.